pigpen-avro/src/main/clojure/pigpen/avro/core.clj

```diff
@@ -0,0 +1,120 @@
+;;     See the License for the specific language governing permissions and
+;;     limitations under the License.
+;;
+;;
+
+(ns pigpen.avro.core
+  (:require [clojure.string :as str]
+            [pigpen.raw :as raw]
+            [pigpen.core :as pig])
+  (:import [org.apache.avro
+            Schema
+            Schema$Field
+            Schema$Type
+            Schema$Parser
+            SchemaNormalization]
+           [org.apache.avro.file DataFileReader]
+           [org.apache.avro.specific SpecificDatumReader]
+           [org.apache.avro.generic GenericData$Record]
+           [pigpen.local PigPenLocalLoader]))
+
+(set! *warn-on-reflection* true)
+
+(declare field-names)
+(defn prefixed-names [record-schema prefix]
+  (for [subfield (field-names record-schema)]
+    (str prefix "." subfield)))
+
+(defn field-names [^Schema parsed-schema]
+ (->> (.getFields parsed-schema)
+       (map (fn [^Schema$Field field]
+              (let [type (.getType ^{:tag Schema} (.schema field))]
+                (cond
+                 (= type Schema$Type/RECORD) (prefixed-names (.schema field) (.name field))
+                 ;; pig only supports union of null, <type>.
+                 (= type Schema$Type/UNION)  (let [record-fields (->> (.getTypes (.schema field))
+                                                                      (filter #(= (.getType ^{:tag Schema} %) Schema$Type/RECORD))
+                                                                      (map (fn [rec] (prefixed-names rec (.name field)))))]
+                                               (if (empty? record-fields) [(.name field)]
+                                                   record-fields))
+                 :else                       (.name field)))))
+       flatten
+       vec))
+
+(defn parse-schema [^String s] (let [parser ^Schema$Parser (Schema$Parser.)]
+                         (.parse parser s)))
+
+(defmethod pigpen.pig/native->clojure org.apache.avro.util.Utf8 [value]
+  (str value))
+
+(defn dotted-keys->nested-map [kvs]
+  (->> kvs
+   (map (fn [[k v]] [(-> k name (clojure.string/split #"\.")) v]))
+   (reduce (fn [acc [ks v]]
+             (try
+              (if v (assoc-in acc (map keyword ks) v) acc)
+              (catch Exception e (throw (Exception. (str "can't assoc-in record: " acc "\nkeys: " ks "\nvalue: " v "\nkvs: " kvs)))))) {})))
+
+
+(defn load-avro
+  "*** ALPHA - Subject to change ***
+
+  Loads data from an avro file. Returns data as maps with keyword keys corresponding to avro field names. Fields with avro type \"map\" will be maps with string keys.
+
+  Example:
+
+    ;; avro schemas are defined [on the project's website](http://avro.apache.org/docs/1.7.7/spec.html#schemas)
+    ;; load-avro takes the schema as a string.
+    (pig-avro/load-avro \"input.avro\" (slurp \"schemafile.json\"))
+
+    (pig-avro/load-avro \"input.avro\"
+       \"{\"namespace\": \"example.avro\", \"type\": \"record\", \"name\": \"foo\",
```

Comment:
In my experience we keep all the schemas in JSON files.

The schema is also available in the data file, as part of the Avro spec. I'd spend the effort to remove the schema argument entirely before defining a DSL.